internal/emlb/emlb.go

```diff
@@ -233,6 +233,113 @@ func (e *EMLB) ReconcileVIPOrigin(ctx context.Context, machineScope *scope.Machi
+
+	// Fetch the Load Balancer object.
+	// Skip if 404, otherwise error
+	lb, err := e.getLoadBalancer(ctx, lbID)
+	if err != nil {
+		log.Error(err, "failed to load the loadbalancer object, cannot proceed with deletion")
+		return err
+	}
+
+	// Get an int version of the listener port number.
+	portNumber, err := strconv.ParseInt(lbPortNumber, 10, 32)
+	if err != nil {
+		log.Error(err, "failed to convert the loadbalancer port number to an int, cannot proceed with deletion")
+		return err
+	}
+
+	// Get the entire listener port object.
+	// Skip if 404, otherwise error
+	lbPort, err := e.getLoadBalancerPort(ctx, lbID, int32(portNumber))
+	if err != nil {
+		log.Error(err, "failed to load the loadbalancer port object, cannot proceed with deletion")
+		return err
+	}
+
+	resp, err := e.deleteListenerPort(ctx, lbPort.GetId())
+	if err != nil {
+		log.Error(err, "LB Port Delete Failed", "EMLB ID", lb.GetId(), "Port ID", lbPort.GetId(), "Response Body", resp.Body)
+		return err
+	}
+
+	resp, err = e.deleteLoadBalancer(ctx, lb.GetId())
+	if err != nil {
+		log.Error(err, "LB Delete Failed", "EMLB ID", lb.GetId(), "Response Body", resp.Body)
+		return err
+	}
+
+	return nil
+}
+
+// DeleteLoadBalancerOrigin deletes the Equinix Metal Load Balancer associated with a given ClusterScope.
+func (e *EMLB) DeleteLoadBalancerOrigin(ctx context.Context, machineScope *scope.MachineScope) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	clusterName := machineScope.Cluster.Name
+
+	// Make sure the machine has an EMLB Pool ID in its packetMachine annotations, otherwise abort.
+	lbPoolID, exists := machineScope.PacketMachine.Annotations[loadBalancerPoolIDAnnotation]
+	if !exists || (lbPoolID == "") {
+		return fmt.Errorf("no Equinix Metal Load Balancer Pool found in machine's annotations")
+	}
+
+	// Make sure the machine has an EMLB Origin ID in its packetMachine annotations, otherwise abort.
+	lbOriginID, exists := machineScope.PacketMachine.Annotations[loadBalancerOriginIDAnnotation]
+	if !exists || (lbOriginID == "") {
+		return fmt.Errorf("no Equinix Metal Load Balancer Origin found in machine's annotations")
+	}
+
+	log.Info("Deleting EMLB Origin from Pool", "Cluster Metro", e.metro, "Cluster Name", clusterName, "Project ID", e.projectID, "Pool ID", lbPoolID, "Origin ID", lbOriginID)
+
+	// Fetch the Load Balancer Pool object.
+	lbPool, err := e.getLoadBalancerPool(ctx, lbPoolID)
+	if err != nil {
+		log.Error(err, "failed to load the loadbalancer pool object, cannot proceed with deletion")
+		return err
+	}
+
+	resp, err := e.deleteOrigin(ctx, lbOriginID)
+	if err != nil {
+		log.Error(err, "LB Origin Delete Failed", "Pool ID", lbPool.GetId(), "Origin ID", lbOriginID, "Response Body", resp.Body)
+		return err
+	}
+
+	// TODO: Validate pool is empty
+
+	log.Info("Deleting EMLB Pool", "Cluster Metro", e.metro, "Cluster Name", clusterName, "Project ID", e.projectID, "Pool ID", lbPoolID)
+
+	resp, err = e.deletePool(ctx, lbPool.GetId())
```

Comment:
I'm not familiar with where this fits in to the overall cluster lifecycle.  Is there any risk here that cluster-api could empty and delete the pool because we're making some kind of cluster change (recreating control plane nodes or something).  Even if there is that risk, does it really matter, or would cluster-api just create a new pool when a new control plane node came up?

Reply:
This is called on node deletion. We have a pool per node so it's no problem, currently.

Reply:
Ah, OK.  If there's a pool per node then I think this is similar to the LB deletion, and deleting the pool will automatically delete the origins inside the pool.

Reply:
Can confirm this works here too and will delete the origin deleting parts. My original idea was to delete all the components explicitly so that if things start failing, as little will be deleted as possible, but given the one to one nature, I guess that doesn't make much sense.